erpnext/manufacturing/doctype/job_card/job_card.py

```diff
@@ -100,6 +108,27 @@ def schedule_time_logs(self, row):
 			self.validate_overlap_for_workstation(args, row)
 			self.check_workstation_time(row)
 
+	def set_expected_and_actual_time(self):
+		# set expected start and end time
+		for doctype, time_fields in {'scheduled_time_logs': ['expected_start_time', 'expected_end_time'],
+			'time_logs': ['actual_start_time', 'actual_end_time']}.items():
+			if not self.get(doctype):
+				continue
+
+			time_list = []
+			for row in self.get(doctype):
+				for field in ['from_time', 'to_time']:
+					if row.get(field):
+						time_list.append(get_datetime(row.get(field)))
+
+			if time_list:
+				self.set(time_fields[0], min(time_list))
+				if time_fields[1] == 'actual_end_time' and not self.time_logs[-1].to_time:
+					self.set(time_fields[1], '')
+					return
+
+				self.set(time_fields[1], max(time_list))
```

Comment:
```suggestion
		for child_table, start_field, end_field in [('scheduled_time_logs', 'expected_start_time', 'expected_end_time'),
			('time_logs', 'actual_start_time', 'actual_end_time')]:
			if not self.get(child_table):
				continue

			time_list = []
			for row in self.get(child_table):
				for field in ['from_time', 'to_time']:
					if row.get(field):
						time_list.append(get_datetime(row.get(field)))

			if time_list:
				self.set(start_field, min(time_list))
				if end_field == 'actual_end_time' and not self.time_logs[-1].to_time:
					self.set(end_field, '')
					return

				self.set(end_field, max(time_list))
```

fields[0] and fields[1] can be reaplced with better variable names. Same with doctype -> child_table.